changes/2025-09-12_introduce-metrics-interface/change.md

```diff
@@ -0,0 +1,169 @@
+This can be seen as a best practice and users may find this a good exercise;
+however, CT's libraries do not make debugging an easy task.
+
+A feature which allows customers to get real-time telemetry of their application's
+integration with CT's libraries would be welcomed by users and will deliver on the
+"easy to use and hard to misuse" tenet.
+
+Introducing a new interface that defines the operations that must be
+implemented in order to build a specification compliant MetricsAgent.
+
+## Requirements
+
+The interface should have three requirements.
+
+1. MUST be simple.
+1. MUST be extensible.
+
+The following is documented to signify its importance
+even though the interface is not able to make this guarantee.
+Every implementation of the proposed interface must
+ensure the following.
+
+1. MUST NOT block the application's execution thread.
+
+## Points of Integration
+
+To collect metrics across CT's library stack multiple points of integration
+are needed in order to collect metrics across CT's libraries.
+Generally, CT's libraries work as follows:
+
+_Note: Not every Client supports the Material Provider Library.
+The Diagram below assumes it to help the mental model._
+
+```mermaid
+sequenceDiagram
+    participant Client
+    box Material Provider Library
+      participant CMM
+      participant Keyring
+    end
+    loop Content Encryption
+        Client->>Client: Content Encryption
+    end
+    Client<<->>CMM: GetEncryption/Decryption Materials
+    CMM<<->>Keyring: OnEncrypt/OnDecrypt
+```
+
+To optionally emit metrics from a top level client,
+all customer facing APIs MUST be changed to optionally accept
+a Metrics Agent. This will allow customers to define and supply one top
+level Metrics Agent; this agent will get plumbed throughout CT's stack.
+
+For example, in the ESDK for Java this would look like:
+
+```java
+final AwsCrypto crypto = AwsCrypto.builder().build();
+// Create a Keyring
+final MaterialProviders matProv =
+  MaterialProviders.builder()
+      .MaterialProvidersConfig(MaterialProvidersConfig.builder().build())
+      .build();
+
+final IKeyring rawAesKeyring = matProv.CreateRawAesKeyring(keyringInput);
+final Map<String, String> encryptionContext =
+    Collections.singletonMap("ExampleContextKey", "ExampleContextValue");
+
+// Create a Metrics Agent
+final IMetricsAgent metrics = matProv.CreateSimpleMetricsAgent(metricsAgentInput);
+// 4. Encrypt the data
+final CryptoResult<byte[], ?> encryptResult =
+    crypto.encryptData(rawAesKeyring, metrics, EXAMPLE_DATA, encryptionContext);
+final byte[] ciphertext = encryptResult.getResult();
+```
+
+This change will allow Crypto Tools to introduce a Metrics Agent in a
+non-breaking way as the Metrics Agent will be an optional parameter
+at customer facing API call sites.
+
+Currently, the ESDK client APIs models are defined [here](https://github.com/aws/aws-encryption-sdk/blob/mainline/AwsEncryptionSDK/dafny/AwsEncryptionSdk/Model/esdk.smithy#L60-L126).
```

Comment:
I assume that the DB-ESDK item encryptor would be similar.
What about the DDB client interface?
Also, are there plans for S3EC?